Frontend/src/Components/Dashboard/GraphContainer.js

```diff
@@ -0,0 +1,253 @@
+ *
+ * @returns a list of categories taken from the JSON file with randomly generated colors
+ */
+function generateCategories() {
+  const categories = [],
+    allDatasets = [];
+  for (const category of GraphData.categories) {
+    const values = category.values.map((obj) => {
+      const colorNum = Math.floor(Math.random() * 0x3fffff + 0x3fffff);
+      const color = "#" + colorNum.toString(16);
+
+      const output = { key: obj.key, name: obj.name, color };
+      allDatasets.push(output);
+      // console.log("Generated color for", obj, ":", output);
+      return output;
+    });
+    categories.push({ category: category.category, values });
+  }
+
+  return [categories, allDatasets];
+}
+
+/**
+ * Creates a new GraphContainer component
+ *
+ * @param {any} props the props to pass to this graph container
+ * @returns the graph-containing component
+ */
+export default function GraphContainer({ queue, latestTime, ...props }) {
+  // fetch constants: categories and flatpacked categories
+  const [categories, allDatasets] = useConst(generateCategories);
+
+  //------------------- Choosing graphs using Select components -------------------
+  // stores the list of graph names/id
+  const [graphTitles, setGraphTitles] = useState(["", "", ""]);
+
+  /**
+   * A function that makes the given graph display in the given place. If the graph is
+   * already being shown in a different place and cannot be duplicated, it will be
+   * swapped with the graph in the given index.
+   *
+   * @param {string} name the name of the graph to show
+   * @param {number} idx the index of the new place the graph should be
+   */
+  const showGraph = useCallback(
+    (name, idx) => {
+      // save graph in old position, if needed
+      // if (graphTitles[idx] && graphTitles[idx] !== "Custom")
+      //   onSave(graphTitles[idx], datasets);
+
+      // duplicate the graph titles
+      const copy = graphTitles.slice();
+
+      // check for special values
+      if (name?.length && name !== "Custom") {
+        const oldIdx = graphTitles.indexOf(name);
+
+        // found old position of graph, swap with new position
+        if (oldIdx !== -1) {
+          // see https://jsbench.me/snl20xx1g9/1 for reasoning behind this algorithm
+          copy[oldIdx] = copy[idx];
+          copy[idx] = name;
+          // console.log(graphTitles, "[", idx, "] =", name, "=", graphTitles);
+
+          setGraphTitles(copy);
+          return;
+        }
+      }
+
+      // default set: if new custom/empty graph or replacing new custom/empty graph
+      // see https://jsbench.me/vsl20xqpso/1 for reasoning behind this algorithm
+      copy[idx] = name;
+      // console.log(graphData, "[", idx, "] =", name, "=", graphData);
+      setGraphTitles(copy);
+    },
+    [graphTitles, setGraphTitles]
+  );
+
+  //------------------------- Saving custom graphs ----------------------------
+  const [customGraphData, setCustomGraphData] = useState({});
```

Comment:
Could you add a quick comment to describe the contents/purpose of customGraphData?

Reply:
Done